src/test/scala/is/hail/methods/SkatSuite.scala

```diff
@@ -0,0 +1,450 @@
+import is.hail.io.annotators.IntervalList
+import is.hail.keytable.KeyTable
+import is.hail.variant.{Genotype, VariantDataset}
+
+import scala.sys.process._
+import breeze.linalg.{qr, _}
+import breeze.numerics.abs
+
+import scala.language.postfixOps
+import is.hail.annotations.Annotation
+import is.hail.stats.{RegressionUtils, SkatModel}
+import org.apache.spark.sql.Row
+import org.testng.annotations.Test
+import org.scalatest.Assertions
+
+
+object SkatAggForR {
+  val zeroValDense = SkatAggForR(new ArrayBuilder[DenseVector[Double]](),new ArrayBuilder[Double])
+  val zeroValSparse = SkatAggForR(new ArrayBuilder[SparseVector[Double]](),new ArrayBuilder[Double])
+
+  def seqOp[T <: Vector[Double]](safr: SkatAggForR[T], info: (T,Double)): SkatAggForR[T] = {
+    val (x,weight) = info
+    SkatAggForR(safr.xs + x, safr.weights + weight)
+  }
+
+  def combOp[T <: Vector[Double]](safr: SkatAggForR[T], safr2: SkatAggForR[T]): SkatAggForR[T] =
+    SkatAggForR(safr.xs ++ safr2.xs.result(), safr.weights ++ safr2.weights.result())
+
+  def sparseResultOP(safr: SkatAggForR[SparseVector[Double]], n: Int): (DenseMatrix[Double], DenseVector[Double]) = {
+    val m = safr.xs.size
+    val xArray = Array.ofDim[Double](m * n)
+
+    var i = 0
+    while (i < m){
+      val index = safr.xs(i).index
+      val data = safr.xs(i).data
+      var j = 0
+      while (j < index.length) {
+        xArray(i * n + index(j)) = data(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    (new DenseMatrix(n, m, xArray), new DenseVector(safr.weights.result()))
+  }
+
+  def denseResultOP(safr: SkatAggForR[DenseVector[Double]], n: Int): (DenseMatrix[Double], DenseVector[Double]) = {
+    val m = safr.xs.size
+    val xArray = Array.ofDim[Double](m * n)
+
+    var i = 0
+    while (i < m){
+      val data = safr.xs(i).data
+      var j = 0
+      while (j < n) {
+        xArray(i * n + j) = data(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    (new DenseMatrix(n, m, xArray), new DenseVector(safr.weights.result()))
+  }
+
+}
+case class SkatAggForR[T <: Vector[Double]](xs: ArrayBuilder[T],weights: ArrayBuilder[Double])
+
+
+class SkatSuite extends SparkSuite {
+
+
+  def testInR(vds: VariantDataset,
+    keyName: String,
+    variantKeys: String,
+    singleKey: Boolean,
+    weightExpr: String,
+    yExpr: String,
+    covExpr: Array[String],
+    hardcalls:Boolean): Array[Row] = {
+
+    if (hardcalls) {
+      polymorphicTestInR(vds, keyName, variantKeys, singleKey, weightExpr, yExpr, covExpr,
+        RegressionUtils.hardCalls(_, _), SkatAggForR.zeroValSparse, SkatAggForR.sparseResultOP _)
+    }
+    else {
+      val dosages = (gs:Iterable[Genotype],n:Int) => RegressionUtils.dosages(gs, (0 until n).toArray)
+      polymorphicTestInR(vds, keyName, variantKeys, singleKey, weightExpr, yExpr, covExpr,
+        dosages, SkatAggForR.zeroValDense, SkatAggForR.denseResultOP _)
+    }
+
+  }
+
+  def polymorphicTestInR[T <: Vector[Double]](vds: VariantDataset,
```

Comment:
I can easily move the code out of the testing suite, when i first wrote it I was just trying to set up the matrix problems easily. I think I can reduce the code a lot because of how we implemented the functions for processing the hardcalls vs dosages. I should be able to just pass in the SkatAggForR class and methods into the function call for the skat routine. I'll mess around cleaning up this code once I get k << n working.